A31-xds-routeaction-support.md

```diff
@@ -0,0 +1,220 @@
+}
+```
+
+The "xds_cluster_manager" policy will maintain a map from cluster name (in each
+`cds_config` entry) to child Policy in order to route Picks.
+
+The Resolver and Config Selector need to coordinate removal of clusters when
+they are deleted by xDS.  In order to prevent deleted clusters from being
+removed from the LB Policy when they may still be needed, the Config Selector
+will mark each chosen cluster as in-use and complete at the start and end of
+each RPC, respectively.  This could be implemented with atomics, locks, or by
+reference-counted objects.  To avoid races, the removal of a cluster will
+require two steps:
+
+1.  The Resolver produces a Config Selector incapable of selecting the deleted
+    cluster, but a Service Config which still contains it.  This will ensure the
+    Load Balancing Policy still contains any cluster chosen by the old Config
+    Selector during the transition.
+1.  After the Config Selector has been updated and the previous one can no
+    longer be called, and when the count of RPCs referencing a deleted cluster
+    is zero, the cluster can safely be removed from the Load Balancing Config in
+    the Service Config.
+
+This wouldn't need to be encoded as an explicit two-step process -- the first
+step should happen as an immediate result of the xDS update, and the second step
+can happen when clusters are no longer referenced by anything.  E.g. if the
+Config Selector has a reference to the deleted cluster, then when there are no
+more references to the config selector, the final reference to the cluster will
+be removed, triggering the Service Config update.
+
+Note that when a cluster is deleted in xDS, if the xds_cluster_manager Load
+Balancing Policy were to queue RPCs in this scenario, wait-for-ready RPCs would
+be stuck indefinitely referencing the cluster, and it could never be removed.
+Deleted clusters need to drop the RPC and return an UNAVAILABLE status.
+
+When new clusters are added to xDS, they will be added to both the Config
+Selector and Service Config simultaneously.  Here, precautions must be taken to
+ensure it is not possible for the Config Selector to be used before the Load
+Balancing Policy's Picker has been updated to be aware of the new cluster.  The
+channel must update the Load Balancing Policy before applying the Config
+Selector, and the Load Balancing Policy must provide a new Picker before its
+Update method returns back to the channel. If the Picker does not recognize a
+cluster name provided to it, which should not be possible when following the
+steps above in that order, it will drop the RPC with an INTERNAL status code.
+
+The following drawing shows which parts of the system will contain which current
+Load Balancing Policies.  The blue box represents the Config Selector and the
+red box represents the top-level Load Balancing Policy and its children.
+
+![xDS Architecture](A31_graphics/xds_architecture.png "xDS Architecture")
```

Comment:
Let's include the graphic in SVG format as well, in case we need to make changes to it later.